Propose new appointment-book hook (#464)
cds-hooks/docs: docs/hooks/appointment-booking.md

# `appointment-booking`

| Metadata | Value
| ---- | ----
| specificationVersion | 1.0

> | hookVersion | 1.0

Reviewer: Is a 1.0 version for this hook appropriate? Regardless, you should also specify a [hookMaturity](https://cds-hooks.org/hooks/template/).

```suggestion
| hookVersion | 1.0
| hookMaturity | 0 - Draft 
```

## Workflow

This hook is invoked when the user is scheduling one or more future encounters/visits for the patient.  It may be invoked at the start and end of the booking process and/or any time between those two points.  It allows hook services to intervene in the decision of when future appointments should be scheduled, where they should be scheduled, what services should be booked, to identify actions that need to occur prior to scheduled appointments, etc.

## Context

> The Patient whose appointment is/appointments are bing booked as well as the draft Appointment records.

Reviewer: ```suggestion
The Patient whose appointment(s) are being booked, as well as the proposed Appointment records.
```

Field | Optionality | Prefetch Token | Type | Description

> ----- | -------- | ---- | ---- | ----

Reviewer: Why not include `userId` in the context (like all of our other hooks) ?

Author: Added

        {
          "actor": {
            "reference": "Practitioner/example",
            "display": "Dr Adam Careful"
          },
          "required": "required",
          "status": "accepted"
        }
      ]
    },
    {
      "resourceType": "Appointment",
      "id": "apt1",
      "status": "proposed",
      "appointmentType": {
        "coding": [
          {
            "system": "http://hl7.org/fhir/v2/0276",
            "code": "CHECKUP",
            "display": "A routine check-up, such as an annual physical"
          }
        ]
      },
      "description": "Regular physical",
      "start": "2020-08-01T13:00:00-06:00",
      "end": "2020-08-01T13:30:00:00-06:00",
      "created": "2019-08-01",
      "participant": [
        {
          "actor": {
            "reference": "Patient/example",
            "display": "Peter James Chalmers"
          },
          "required": "required",
          "status": "tentative"
        },
        {
          "actor": {
            "reference": "Practitioner/example",
            "display": "Dr Adam Careful"
          },
          "required": "required",
          "status": "accepted"
        }
      ]
    }
  ]
}
```

> ```json

Reviewer: Is this the R4 example? 

Could you please label each example with the FHIR version it implements?

Author: Both are STU3.  Label added

# `appointment-booking`

| Metadata | Value
| ---- | ----
| specificationVersion | 1.0
| hookVersion | 1.0
| hookMaturity | 0 - Draft 

## Workflow

> This hook is invoked when the user is scheduling one or more future encounters/visits for the patient.  It may be invoked at the start and end of the booking process and/or any time between those two points.  It allows hook services to intervene in the decision of when future appointments should be scheduled, where they should be scheduled, what services should be booked, to identify actions that need to occur prior to scheduled appointments, etc.

Reviewer 2: ```suggestion
This hook is invoked when the user is scheduling one or more future encounters/visits for the patient.  It may be invoked at the start and end of the booking process and/or any time between those two points.  It allows CDS Services to intervene in the decision of when future appointments should be scheduled, where they should be scheduled, what services should be booked, to identify actions that need to occur prior to scheduled appointments, etc.
```

'CDS Services', not 'hook services'
